Non-affine Piola and hanging nodes support for matrix-free RT

NiklasWik opened the pull request:
Adds support for non-affine meshes and adds test for hanging nodes for the matrix-free implementation of `FE_RaviartThomasNodal`. 

This adds a more data to `mapping_info_storage` which is not ideal. To avoid this one would have to change how the Hessians are computed, i.e to compute the inverse jacobian gradient on the fly. 

Still missing for the matrix-free evaluation of RT is support for cells in non-standard orientation.

peterrum commented on tests/matrix_free/matrix_vector_rt_face_03.cc:
```diff
@@ -0,0 +1,71 @@
+  FE_RaviartThomasNodal<dim> fe(fe_degree - 1);
+  DoFHandler<dim>            dof(tria);
+  dof.distribute_dofs(fe);
+  AffineConstraints<double> constraints;
+  constraints.close();
```
Don't you need to apply any constraints? The normal component is continuous, isn't it?

NiklasWik replied:
Clearly missed some parts hehe :') But I think I have them all now..?

peterrum commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -3139,16 +3137,14 @@ namespace internal
       std::integral_constant<bool, false>)
     {
-      // TODO. Need to make sure hanging nodes work, i.e call
-      // create_eval_tensor_product with correct direction.
```
Is it really enough to treat the hanging nodes only by `FEFaceEvaluation`? What does happen you use `FEEvaluation`? The contributions of constrained DoFs are not added to the constraining ones.

NiklasWik replied:
The shape data will be the same for `FEEvaluation` since each cell is still evaluated in the entire reference cell. For `FEFaceEvaluation` we instead have to evaluate either between coordinates `[0, 0.5]x[0, 0.5]`, or `[0, 0.5]x[0.5, 1]`, etc, in 3D. It's `[0, 0.5]` or `[0.5, 1]` in 2D.

As far as I know, the constrained dofs are just ignored when calling `FEEvaluation.distribute_local_to_global(..)`, and only updated if you call `constraints.distribute(..)`, but you probably know this a lot better than me.

peterrum replied:
> As far as I know, the constrained dofs are just ignored when calling FEEvaluation.distribute_local_to_global(..), and only updated if you call constraints.distribute(..), but you probably know this a lot better than me.

They are not ignored, but their contributions are added to the constraining DoFs. Your test should give you different results depending on `DoFTools::make_hanging_node_constraints` was called or not? Could you check?

NiklasWik replied:
> Your test should give you different results depending on `DoFTools::make_hanging_node_constraints` was called or not?

It does not make a difference if I call it or not. Maybe I need a different test? I also does not manage to finish on the CI check since the matrix assembly is so damn slow for some reason.

peterrum replied:
> It does not make a difference if I call it or not. Maybe I need a different test?

There should be a difference. At the moment, you are only comparing matrix-free and matrix-based results. Those should be consistent if correctly implemented. If you add hanging-node constraints, the l2 norm of the vectors should differ.

NiklasWik replied:
Ah, yes ofc. It does make a difference!

peterrum commented on tests/matrix_free/matrix_vector_rt_face_common.h:
```diff
@@ -308,18 +309,21 @@ do_test(const DoFHandler<dim> &          dof,
   system_matrix.vmult(ref, initial_condition);
 
+  // zero constrained dofs
+  for (unsigned int i = 0; i < dof.n_dofs(); ++i)
+    if (constraints.is_constrained(i))
+      ref(i) = 0;
```
```suggestion
  constraints.zero_out(ref);
```

peterrum replied:
It is called `set_zero()`. Sorry for the confusion.

peterrum reviewed (commented):
Looks good!

peterrum commented:
/rebuild

NiklasWik commented:
What kind of test would you recommend me writing instead @peterrum @kronbichler? Some convergence test maybe?

peterrum commented:
> What kind of test would you recommend me writing instead @peterrum @kronbichler? Some convergence test maybe?

Yes, that sounds reasonable! In addition, I would extend the existing tests so that they also output the l2-norms of either matrix-based or the matrix-free solution!

peterrum commented:
```
770/770 Test #1650: matrix_free/matrix_vector_rt_face_03.debug ............................................***Timeout 600.00 sec
```
Amazing :D

NiklasWik commented:
> Yes, that sounds reasonable! 

Do you know of any examples at the top of your head that I can use as a base? I would like to avoid writing something completely from scratch hehe :)

peterrum commented:
@NiklasWik Could you rebase?

NiklasWik commented:
You might have preferred a new PR but I just instead renamed this one and changed the first comment. Hope that's ok! :)

I added the l2-norm as you suggested @peterrum, but the output changed after I rebased so I ended up removing it anyway. 

I also lowered the number of cells for the hanging node test. Let's see if it finishes in time now! (Also saw that @kronbichler optimized the tests, which have been enough)
